### PR TITLE
SONARJAVA-1715 rule lookup type hierarchy

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 1]
```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
...
-        return false;
-      }
-      if (isNotObjectMethod((MethodTree) member) && isNonStaticNonDefaultMethod(member)) {
-        methods++;
-      }
+    List<Tree> members = classTree.members();
+    if (members.stream().allMatch(tree -> tree.is(Tree.Kind.METHOD))) {
+      return members.stream()
+        .map(tree -> (MethodTree) tree)
+        .filter(method -> isNotObjectMethod(method) && isNonStaticNonDefaultMethod(method))
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 2]
```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
...
+        .count() == 1;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 3]
```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
...
-    ImmutableCollection<List<String>> methods = OBJECT_METHODS.get(method.simpleName().name());
-    if (methods != null) {
-      for (List<String> types : methods) {
-        if (sameParameters(method.parameters(), types)) {
-          return false;
-        }
-      }
-    }
-    return true;
+  private static boolean interfaceIsNotEmpty(ClassJavaType type) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Private method 'interfaceIsNotEmpty' is never used. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 4]
```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
...
-    if (provided.size() == args.size()) {
-      for (VariableTree var : provided) {
-        args.remove(var.type().symbolType().name());
-      }
-      if (args.isEmpty()) {
-        return true;
-      }
-    }
-    return false;
+  private static boolean isNotThisOrSuper(Symbol symbol) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Private method 'isNotThisOrSuper' is never used. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 6 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 5]
```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
...
-    for (Tree member : classTree.members()) {
-      boolean isMethod = member.is(Tree.Kind.METHOD);
-      if (!isMethod) {
-        return false;
-      }
-      if (isNotObjectMethod((MethodTree) member) && isNonStaticNonDefaultMethod(member)) {
-        methods++;
-      }
+    List<Tree> members = classTree.members();
+    if (members.stream().allMatch(tree -> tree.is(Tree.Kind.METHOD))) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: This is not correct : A functional interfaces can have constants defined. or inner classes.

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 6]
```diff
@@ -78,70 +78,55 @@ public void visitNode(Tree tree) {
...
+    if (!symbol.isTypeSymbol()) {
+      // unknown interface
+      return Integer.MAX_VALUE;
     }
-    return false;
+    Symbol.TypeSymbol interfaceSymbol = (Symbol.TypeSymbol) symbol;
+    return interfaceSymbol.memberSymbols().stream()
+      .filter(member -> member.isMethodSymbol())
+      .filter(member -> {
+        Symbol.MethodSymbol methodSymbol = (Symbol.MethodSymbol) member;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Make this line start at column 11. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

File: java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java [hunk 7]
```diff
@@ -78,70 +78,55 @@ public void visitNode(Tree tree) {
...
-        if (sameParameters(method.parameters(), types)) {
-          return false;
-        }
-      }
-    }
-    return true;
+  private static boolean noAbstractMethodInParentInterfaces(List<Type> interfaces) {
+    return interfaces.stream()
+      .map(Type::symbol)
+      .noneMatch(symbol -> numberOfAbstractMethod(symbol) > 0 || !noAbstractMethodInParentInterfaces(symbol.interfaces()));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

**Pull request merged**